# <span style='color:Blue;align:center'><center>Pune Air Quality Index Analysis - Data Cleaning</center></span>
### <span style='color:Orange'>Does air quality index worsen during festival season?</span>
<i>Note : Pune Municipal Corporation cannot warrant the quality or accuracy of the data.</i>

<a id="main"></a>
<div class="alert alert-block alert-success">
    <h2>Index</h2>
</div>

1. [Read and Clean Dataframe Structure](#1.datacleaning)
2. [Feature Cleaning](#2.featurecleaning)


In [65]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import re

In [2]:
# Read list of files to be read
!ls

Air Quality Parameters - MPCB (2006-08).xlsx
Air Quality Parameters - MPCB (2009-11).xlsx
Air Quality Parameters - MPCB (2012-14).xlsx
Air Quality Parameters - MPCB (2015-17).xlsx
Air Quality Parameters - MPCB (2018-19 March).xlsx
airpollution_data_cleaning.ipynb
cleaned.csv
dates
dates.csv
library_functions.ipynb


<a id="1.datacleaning"></a>
<div class="alert alert-block alert-info">
    <h2>1. Read and Clean Dataframe Structure</h2>
</div>

   1. [Read Dataframe](#1.1.dataframe)
   2. [Dataframe Cleaning](#1.2.dfcleaning)

[go to index](#main)

<a id="1.1.dataframe"></a>

### Read 5 dataframes for data from 2006 till 2019 March

[go to sub heading](#1.datacleaning)

In [3]:
air_df_6_8 = pd.read_excel('Air Quality Parameters - MPCB (2006-08).xlsx')
air_df_9_11 = pd.read_excel('Air Quality Parameters - MPCB (2009-11).xlsx')
air_df_12_14 = pd.read_excel('Air Quality Parameters - MPCB (2012-14).xlsx')
air_df_15_17 = pd.read_excel('Air Quality Parameters - MPCB (2015-17).xlsx')
air_df_18_19 = pd.read_excel('Air Quality Parameters - MPCB (2018-19 March).xlsx')

In [4]:
air_df_6_8.head(1)
#it is observed that columns after "Location" are got created in error and are blank.

Sr.No.        Date SO2 µg/m3 Nox µg/m3  RSPM µg/m3  SPM  CO2 µg/m3   AQI  \
0       1  19-06-2007         6        46        49.0  NaN        NaN  58.0   

  Location  Unnamed: 9  Unnamed: 10  Unnamed: 11 Unnamed: 12  
0  MPCB-KR         NaN          NaN          NaN         NaN

In [5]:
air_df_9_11.head(1)

Sr.No.                 Date SO2 µg/m3 Nox µg/m3  RSPM µg/m3  SPM  \
0       1  2009-01-01 00:00:00        15        53       179.0  NaN   

   CO2 µg/m3  AQI Location  
0        NaN  153  MPCB-KR

In [6]:
air_df_12_14.head(3)

Standards           Unnamed: 1     80      80.1    100 Unnamed: 5    ---  \
0    Sr.No.                 Date    SO2       NOx   RSPM        NaN    CO2   
1       NaN                  NaN  µg/m3     µg/m3  µg/m3        SPM  µg/m3   
2         1  2012-01-01 00:00:00     10  BDL - NA    150        NaN    NaN   

  Unnamed: 7  Unnamed: 8  
0        AQI    Location  
1        NaN         NaN  
2        133  Karve Road

In [7]:
# The column structure is distorted in this dataframe, hence promoting first row to column
air_df_12_14.columns = air_df_12_14.iloc[0]

In [8]:
# Read the data from 3rd Row as actual data starts from there
air_df_12_14 = air_df_12_14.iloc[2:]

In [9]:
air_df_12_14.head(1)

0 Sr.No.                 Date SO2       NOx RSPM  NaN  CO2  AQI    Location
2      1  2012-01-01 00:00:00  10  BDL - NA  150  NaN  NaN  133  Karve Road

In [10]:
air_df_15_17.head(1)
#need to remove blank columns at end

Sr.No.                 Date  SO2 µg/m3 Nox µg/m3  RSPM µg/m3  SPM  \
0       1  2015-01-01 00:00:00         19         9         198  NaN   

   CO2 µg/m3  AQI    Location  Unnamed: 9  Unnamed: 10  
0        NaN  165  Karve Road         NaN          NaN

In [11]:
air_df_18_19.head(1)

Sr No                 Date  SO2 µg/m3  Nox µg/m3  RSPM µg/m3    SPM  \
0      1  2018-02-01 00:00:00         49         87         263  376.0   

   CO2 µg/m3    AQI   Location  
0        NaN  213.0  MPCB-BSRI

<a id="1.2.dfcleaning"></a>

### Clean the dataframes

[go to sub heading](#1.datacleaning)

In [12]:
def dropcolumnsaftercolumn(df,column_name):
    '''
    This function returns a dataframe after dropping all columns after a specific column.
    Parameters
    ------
        df - Input Dataframe
        column_name - Name of the column after which all columns needs to be dropped

    Returns
    ------
        Modified dataframe
    '''
    last = list(df.columns).index(column_name) + 1
    return df.drop(columns=list(df.columns[last:]),axis=1)

In [13]:
# Remove the columns which are blank and present after location column which is the last column having data
air_df_6_8 = dropcolumnsaftercolumn(air_df_6_8,'Location')
air_df_9_11 = dropcolumnsaftercolumn(air_df_9_11,'Location')
air_df_12_14 = dropcolumnsaftercolumn(air_df_12_14,'Location')
air_df_15_17 = dropcolumnsaftercolumn(air_df_15_17,'Location')
air_df_18_19 = dropcolumnsaftercolumn(air_df_18_19,'Location')

In [14]:
#Check if column sequence is same for all the files and after that we will join the dataframes

In [15]:
air_df_6_8.head(1)

Sr.No.        Date SO2 µg/m3 Nox µg/m3  RSPM µg/m3  SPM  CO2 µg/m3   AQI  \
0       1  19-06-2007         6        46        49.0  NaN        NaN  58.0   

  Location  
0  MPCB-KR

In [16]:
air_df_9_11.head(1)

Sr.No.                 Date SO2 µg/m3 Nox µg/m3  RSPM µg/m3  SPM  \
0       1  2009-01-01 00:00:00        15        53       179.0  NaN   

   CO2 µg/m3  AQI Location  
0        NaN  153  MPCB-KR

In [17]:
air_df_12_14.head(1)

0 Sr.No.                 Date SO2       NOx RSPM  NaN  CO2  AQI    Location
2      1  2012-01-01 00:00:00  10  BDL - NA  150  NaN  NaN  133  Karve Road

In [18]:
air_df_15_17.head(1)

Sr.No.                 Date  SO2 µg/m3 Nox µg/m3  RSPM µg/m3  SPM  \
0       1  2015-01-01 00:00:00         19         9         198  NaN   

   CO2 µg/m3  AQI    Location  
0        NaN  165  Karve Road

In [19]:
air_df_18_19.head(1)

Sr No                 Date  SO2 µg/m3  Nox µg/m3  RSPM µg/m3    SPM  \
0      1  2018-02-01 00:00:00         49         87         263  376.0   

   CO2 µg/m3    AQI   Location  
0        NaN  213.0  MPCB-BSRI

In [20]:
# Lets simplyfy the column names by taking first word and removing others. The same column names will be used 
# for all dataframes
columns_names = [column.lower().split(' ')[0] for column in air_df_6_8.columns]

In [21]:
columns_names

['sr.no.', 'date', 'so2', 'nox', 'rspm', 'spm', 'co2', 'aqi', 'location']

In [22]:
air_df_6_8.columns = columns_names
air_df_9_11.columns = columns_names
air_df_12_14.columns = columns_names
air_df_15_17.columns = columns_names
air_df_18_19.columns = columns_names

In [23]:
#concat the dataframes along axis 0
df = pd.concat([air_df_6_8 , air_df_9_11,air_df_12_14,air_df_15_17,air_df_18_19],ignore_index=True)

<a id="2.featurecleaning"></a>
<div class="alert alert-block alert-info">
    <h2>2. Read and Clean Features</h2>
</div>

   1. [Location](#2.1.location)
   2. [SO2](#2.2.so2)
   3. [nox](#2.3.nox)
   4. [RSPM](#2.4.rspm)
   5. [SPM](#2.5.spm)
   6. [CO2](#2.6.co2)
   7. [AQI](#2.7.aqi)
   8. [Date](#2.8.date)

[go to index](#main)

<a id="2.1.location"></a>

### 2.1. Location

In [24]:
# Lets explore location columns
df['location'].value_counts()

Karve Road          2069
MPCB-KR             1825
MPCB-PMPR           1162
Pimpri Chinchwad     854
MPCB-NS              707
MPCB-BSRI            622
Bhosari              605
MPCB-SWGT            604
Swargate             587
Nal Stop             575
Pimpri               254
Name: location, dtype: int64

In [25]:
# It seems there are two variations of the same location e.g. Swargate and MPCB-SWGT are same. Similar with Nal Stop
# and MPCB-NS, Lets combine them
df['location'].replace(['MPCB-KR','MPCB-PMPR','MPCB-NS','MPCB-BSRI','MPCB-SWGT'],
                       ['Karve Road','Pimpri Chinchwad','Nal Stop','Bhosari','Swargate'],inplace=True)

In [26]:
df['location'].value_counts()

Karve Road          3894
Pimpri Chinchwad    2016
Nal Stop            1282
Bhosari             1227
Swargate            1191
Pimpri               254
Name: location, dtype: int64

In [27]:
df.head(50)
# there seems to be many columns which have non numeric values. Let's explore them in details

sr.no.                 date      so2 nox rspm  spm  co2 aqi    location
0       1           19-06-2007        6  46   49  NaN  NaN  58  Karve Road
1       2           20-06-2007        6  43   29  NaN  NaN  54  Karve Road
2       3           21-06-2007        5  43   26  NaN  NaN  54  Karve Road
3       4           23-06-2007        7  46   79  NaN  NaN  79  Karve Road
4       5           24-06-2007        5  45   19  NaN  NaN  56  Karve Road
5       6           25-06-2007        6  45   42  NaN  NaN  56  Karve Road
6       7           26-06-2007        4  43   64  NaN  NaN  64  Karve Road
7       8           27-06-2007        4  43   53  NaN  NaN  54  Karve Road
8       9           28-06-2007        4  44   47  NaN  NaN  55  Karve Road
9      10           30-06-2007        4  46   27  NaN  NaN  58  Karve Road
10     11  2007-02-07 00:00:00        4  43   44  NaN  NaN  54  Karve Road
11     12  2007-03-07 00:00:00        4  47   62  NaN  NaN  62  Karve Road
12     13  2007-04-07 00:00:00        4  46   62  NaN  NaN  62  Karve Road
13     14  2007-05-07 00:00:00        4  43   55  NaN  NaN  55  Karve Road
14     15  2007-06-07 00:00:00        4  18   84  NaN  NaN  84  Karve Road
15     16  2007-07-07 00:00:00        4  20   79  NaN  NaN  79  Karve Road
16     17  2007-08-07 00:00:00        4  19   63  NaN  NaN  63  Karve Road
17     18  2007-10-07 00:00:00        4  22   73  NaN  NaN  73  Karve Road
18     19  2007-11-07 00:00:00        4  23   67  NaN  NaN  67  Karve Road
19     20  2007-12-07 00:00:00        4  24   66  NaN  NaN  66  Karve Road
20     21           13-07-2007        4  24   53  NaN  NaN  53  Karve Road
21     22           14-07-2007        4  27   58  NaN  NaN  58  Karve Road
22     23           15-07-2007  BDL - 3  21   65  NaN  NaN  65  Karve Road
23     24           16-07-2007        4  22   78  NaN  NaN  78  Karve Road
24     25           17-07-2007        4  22   65  NaN  NaN  65  Karve Road
25     26           18-07-2007        4  26   58  NaN  NaN  58  Karve Road
26     27           19-07-2007  BDL - 3  18   84  NaN  NaN  84  Karve Road
27     28           20-07-2007  BDL - 3  21   65  NaN  NaN  65  Karve Road
28     29           21-07-2007       34  46   69  NaN  NaN  69  Karve Road
29     30           22-07-2007  BDL - 3  46   73  NaN  NaN  73  Karve Road
30     31           23-07-2007        4  20   78  NaN  NaN  78  Karve Road
31     32           24-07-2007        5  43   57  NaN  NaN  57  Karve Road
32     33           25-07-2007        5  46   58  NaN  NaN  58  Karve Road
33     34           26-07-2007        5  43   45  NaN  NaN  54  Karve Road
34     35           27-07-2007        5  43   31  NaN  NaN  54  Karve Road
35     36           28-07-2007        5  24   31  NaN  NaN  31  Karve Road
36     37           29-07-2007        5  32   35  NaN  NaN  40  Karve Road
37     38           30-07-2007        5  30   44  NaN  NaN  44  Karve Road
38     39           31-07-2007  BDL - 3  45   56  NaN  NaN  56  Karve Road
39     40  2007-01-08 00:00:00        5  33   38  NaN  NaN  41  Karve Road
40     41  2007-02-08 00:00:00        5  29   36  NaN  NaN  36  Karve Road
41     42  2007-03-08 00:00:00        5  32   28  NaN  NaN  40  Karve Road
42     43  2007-04-08 00:00:00  BDL - 3  42   27  NaN  NaN  53  Karve Road
43     44  2007-05-08 00:00:00  BDL - 3  33   33  NaN  NaN  41  Karve Road
44     45  2007-06-08 00:00:00  BDL - 3  28   39  NaN  NaN  39  Karve Road
45     46  2007-07-08 00:00:00        9  56   31  NaN  NaN  70  Karve Road
46     47  2007-08-08 00:00:00        4  42   28  NaN  NaN  53  Karve Road
47     48  2007-09-08 00:00:00        5  43   57  NaN  NaN  57  Karve Road
48     49  2007-10-08 00:00:00  BDL - 3  33   48  NaN  NaN  48  Karve Road
49     50  2007-11-08 00:00:00  BDL - 2  33   40  NaN  NaN  41  Karve Road

<a id="2.2.so2"></a>
### 2.2. SO2

[Return to sub heading](#2.featurecleaning)

In [28]:
df[df['so2'].apply(lambda a: type(a) is not int)]['so2'].value_counts()

BDL - 3     50
BDL - 2     48
BDL - NA    40
BDL - NA    13
BDL - 1      2
Name: so2, dtype: int64

<b>BDL</b> means below detection limit. Hence we can make them 0 for BDL - NA and keep the digit for other values

In [29]:
def columnvaluecleaning(series,conversion_type: np.dtype):
    '''
    This function transfors a pandas series by removing non-number chars from string, replacing
    blank with 0 and nan with 0.
    The series is converted to conversion type (dtype)
    Parameters
    ------
        df - Input Dataframe
        column_name - Name of the column after which all columns needs to be dropped

    Returns
    ------
        Modified dataframe
    '''
    #remove everything apart from numbers and dot (int or float)
    series = series.replace('[^0-9.]','',regex=True) #whatever which is not number of dot, replace with blank
    series = series.replace('',0)
    series = series.fillna(0)
    return series.astype(conversion_type)

df['so2'] = columnvaluecleaning(df['so2'],np.int64)

<a id="2.3.nox"></a>

### 2.3. nox

[return to sub heading](#2.featurecleaning)

In [30]:
df[df['nox'].apply(lambda a: type(a) is not int)]['nox'].value_counts()

BDL - NA    77
BDL - NA    60
BDL - 8     17
BDL - 7     14
BDL - 6     12
BDL - 6      2
BDL - 5      2
BDL - 8      1
BDL - 2      1
Name: nox, dtype: int64

In [31]:
df['nox'] = columnvaluecleaning(df['nox'],np.int64)

In [32]:
df[df['nox'].apply(lambda a: type(a) is not int)]['nox'].value_counts()

Series([], Name: nox, dtype: int64)

<a id="2.4.rspm"></a>

### 2.4. RSPM

[return to sub heading](#2.featurecleaning)

In [33]:
df[df['rspm'].apply(lambda a: type(a) is not int)]['rspm'].value_counts()

65.0     39
50.0     39
41.0     39
73.0     37
55.0     36
         ..
228.0     1
226.0     1
225.0     1
223.0     1
11.0      1
Name: rspm, Length: 262, dtype: int64

In [34]:
df['rspm'] = columnvaluecleaning(df['rspm'],np.int64)

In [35]:
df[df['rspm'].apply(lambda a:type(a) is not int)]['rspm'].value_counts()

Series([], Name: rspm, dtype: int64)

<a id="2.5.spm"></a>

### 2.5. SPM

[return to subheading](#2.featurecleaning)

In [36]:
df[df['spm'].apply(lambda a:type(a) is not int)]['spm'].value_counts()

220.0    14
138.0    14
239.0    13
276.0    13
257.0    13
         ..
464.0     1
462.0     1
461.0     1
459.0     1
0.0       1
Name: spm, Length: 506, dtype: int64

In [37]:
df['spm'] = columnvaluecleaning(df['spm'],np.int64)

In [38]:
df[df['spm'].apply(lambda a:type(a) is not int)]['spm'].value_counts()

Series([], Name: spm, dtype: int64)

<a id="2.6.co2"></a>

### 2.6. CO2

[return to subheading](#2.featurecleaning)

In [39]:
df[df['co2'].apply(lambda a:type(a) is not int)]['co2'].value_counts()

121.0    4
134.0    3
145.0    3
150.0    3
152.0    3
        ..
129.0    1
124.0    1
218.0    1
114.0    1
25.0     1
Name: co2, Length: 67, dtype: int64

In [40]:
df['co2'] = columnvaluecleaning(df['co2'],np.int64)

In [41]:
df[df['co2'].apply(lambda a:type(a) is not int)]['co2'].value_counts()

Series([], Name: co2, dtype: int64)

<a id="2.7.aqi"></a>
### AQI

[return to submenu](#2.featurecleaning)

In [42]:
df[df['aqi'].apply(lambda a:type(a) is not int)]['aqi'].value_counts()

123.0    55
119.0    52
117.0    50
113.0    47
115.0    45
         ..
228.0     1
230.0     1
233.0     1
236.0     1
444.0     1
Name: aqi, Length: 206, dtype: int64

In [43]:
df['aqi'] = columnvaluecleaning(df['aqi'],np.int64)

In [44]:
df[df['aqi'].apply(lambda a:type(a) is not int)]['aqi'].value_counts()

Series([], Name: aqi, dtype: int64)

In [45]:
df.head(50)

sr.no.                 date  so2  nox  rspm  spm  co2  aqi    location
0       1           19-06-2007    6   46    49    0    0   58  Karve Road
1       2           20-06-2007    6   43    29    0    0   54  Karve Road
2       3           21-06-2007    5   43    26    0    0   54  Karve Road
3       4           23-06-2007    7   46    79    0    0   79  Karve Road
4       5           24-06-2007    5   45    19    0    0   56  Karve Road
5       6           25-06-2007    6   45    42    0    0   56  Karve Road
6       7           26-06-2007    4   43    64    0    0   64  Karve Road
7       8           27-06-2007    4   43    53    0    0   54  Karve Road
8       9           28-06-2007    4   44    47    0    0   55  Karve Road
9      10           30-06-2007    4   46    27    0    0   58  Karve Road
10     11  2007-02-07 00:00:00    4   43    44    0    0   54  Karve Road
11     12  2007-03-07 00:00:00    4   47    62    0    0   62  Karve Road
12     13  2007-04-07 00:00:00    4   46    62    0    0   62  Karve Road
13     14  2007-05-07 00:00:00    4   43    55    0    0   55  Karve Road
14     15  2007-06-07 00:00:00    4   18    84    0    0   84  Karve Road
15     16  2007-07-07 00:00:00    4   20    79    0    0   79  Karve Road
16     17  2007-08-07 00:00:00    4   19    63    0    0   63  Karve Road
17     18  2007-10-07 00:00:00    4   22    73    0    0   73  Karve Road
18     19  2007-11-07 00:00:00    4   23    67    0    0   67  Karve Road
19     20  2007-12-07 00:00:00    4   24    66    0    0   66  Karve Road
20     21           13-07-2007    4   24    53    0    0   53  Karve Road
21     22           14-07-2007    4   27    58    0    0   58  Karve Road
22     23           15-07-2007    3   21    65    0    0   65  Karve Road
23     24           16-07-2007    4   22    78    0    0   78  Karve Road
24     25           17-07-2007    4   22    65    0    0   65  Karve Road
25     26           18-07-2007    4   26    58    0    0   58  Karve Road
26     27           19-07-2007    3   18    84    0    0   84  Karve Road
27     28           20-07-2007    3   21    65    0    0   65  Karve Road
28     29           21-07-2007   34   46    69    0    0   69  Karve Road
29     30           22-07-2007    3   46    73    0    0   73  Karve Road
30     31           23-07-2007    4   20    78    0    0   78  Karve Road
31     32           24-07-2007    5   43    57    0    0   57  Karve Road
32     33           25-07-2007    5   46    58    0    0   58  Karve Road
33     34           26-07-2007    5   43    45    0    0   54  Karve Road
34     35           27-07-2007    5   43    31    0    0   54  Karve Road
35     36           28-07-2007    5   24    31    0    0   31  Karve Road
36     37           29-07-2007    5   32    35    0    0   40  Karve Road
37     38           30-07-2007    5   30    44    0    0   44  Karve Road
38     39           31-07-2007    3   45    56    0    0   56  Karve Road
39     40  2007-01-08 00:00:00    5   33    38    0    0   41  Karve Road
40     41  2007-02-08 00:00:00    5   29    36    0    0   36  Karve Road
41     42  2007-03-08 00:00:00    5   32    28    0    0   40  Karve Road
42     43  2007-04-08 00:00:00    3   42    27    0    0   53  Karve Road
43     44  2007-05-08 00:00:00    3   33    33    0    0   41  Karve Road
44     45  2007-06-08 00:00:00    3   28    39    0    0   39  Karve Road
45     46  2007-07-08 00:00:00    9   56    31    0    0   70  Karve Road
46     47  2007-08-08 00:00:00    4   42    28    0    0   53  Karve Road
47     48  2007-09-08 00:00:00    5   43    57    0    0   57  Karve Road
48     49  2007-10-08 00:00:00    3   33    48    0    0   48  Karve Road
49     50  2007-11-08 00:00:00    2   33    40    0    0   41  Karve Road

In [46]:
df.tail(50)

sr.no.                 date  so2  nox  rspm  spm  co2  aqi  location
9814     24  2018-04-04 00:00:00   23   67    84  310    0   84  Swargate
9815     25  2018-07-04 00:00:00   25   51    80  288    0   80  Swargate
9816     26  2018-11-04 00:00:00   10   94   119  310    0  114  Swargate
9817     27           18-04-2018   40   60    98  206    0   98  Swargate
9818     28  2018-02-06 00:00:00   36   86    54  136    0  106  Swargate
9819     29  2018-06-06 00:00:00   41   71    47  187    0   89  Swargate
9820     30  2018-09-06 00:00:00   25   47    18  149    0   59  Swargate
9821     31           13-06-2018   27   44    42  123    0   55  Swargate
9822     32           20-06-2018   33   59    45  141    0   74  Swargate
9823     33           22-06-2018   40   64    91  156    0   91  Swargate
9824     34           27-06-2018   32   39    28  102    0   49  Swargate
9825     35           30-06-2018   44   64    32  148    0   80  Swargate
9826     36  2018-04-07 00:00:00   28   88    18   50    0  108  Swargate
9827     37  2018-07-07 00:00:00   14   39    34   49    0   49  Swargate
9828     38           14-07-2018   44   67    27   47    0   84  Swargate
9829     39           18-07-2018   42   45    18   30    0   56  Swargate
9830     40           21-07-2018   47   57    15   28    0   71  Swargate
9831     41           25-07-2018   32   44    26   49    0   55  Swargate
9832     42           28-07-2018   30   45    13   34    0   56  Swargate
9833     43           31-08-2018   53   72    14  148    0   90  Swargate
9834     44  2018-05-09 00:00:00   73   65    13   81    0   91  Swargate
9835     45  2018-08-09 00:00:00   52   64    37  164    0   80  Swargate
9836     46  2018-12-09 00:00:00   42   32    32  109    0   53  Swargate
9837     47           26-09-2018   44   66    47   59    0   83  Swargate
9838     48           29-09-2018   49   56    69  154    0   70  Swargate
9839     49  2018-03-10 00:00:00   46  100    50  150    0  120  Swargate
9840     50  2018-07-10 00:00:00   43   66    89  314    0   89  Swargate
9841     51  2018-10-10 00:00:00   32  112    84  292    0  132  Swargate
9842     52           13-10-2018   42  102    85  295    0  122  Swargate
9843     53           29-10-2018   48  146   103  221    0  166  Swargate
9844     54  2018-01-11 00:00:00   50  110   136  340    0  130  Swargate
9845     55  2018-02-11 00:00:00   51   46   115  180    0  110  Swargate
9846     56  2018-03-11 00:00:00   59   97    75  240    0  117  Swargate
9847     57           29-11-2018   44  155   158  276    0  175  Swargate
9848     58  2018-01-12 00:00:00   46   84   166  301    0  144  Swargate
9849     59  2018-05-12 00:00:00   53  108   142  342    0  128  Swargate
9850     60  2018-08-12 00:00:00   33  106    95  296    0  126  Swargate
9851     61  2018-12-12 00:00:00   40  102    94  252    0  122  Swargate
9852     62           15-12-2018   33   71   145  347    0  130  Swargate
9853     63           16-12-2018   35  118   176  345    0  151  Swargate
9854     64           18-12-2018   49  148   156  369    0  168  Swargate
9855     65           25-12-2018   49  167   189  350    0  187  Swargate
9856     66           29-12-2018   39  108   179  551    0  153  Swargate
9857      1  2019-01-01 00:00:00   31   78   163  362    0  142  Swargate
9858      2  2019-12-01 00:00:00   37  103   179  444    0  153  Swargate
9859      3           18-01-2019   41  100   134  300    0  123  Swargate
9860      4           19-01-2019   50  132   172  446    0  152  Swargate
9861      5           23-01-2019   47  138   210  558    0  173  Swargate
9862      6  2019-02-02 00:00:00   29   88    82  259    0  108  Swargate
9863      7  2019-09-02 00:00:00   39   96   171  422    0  147  Swargate

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9864 entries, 0 to 9863
Data columns (total 9 columns):
sr.no.      9864 non-null object
date        9864 non-null object
so2         9864 non-null int64
nox         9864 non-null int64
rspm        9864 non-null int64
spm         9864 non-null int64
co2         9864 non-null int64
aqi         9864 non-null int64
location    9864 non-null object
dtypes: int64(6), object(3)
memory usage: 693.7+ KB


<a id="2.8.date"></a>
### Date
[return to subheading](#2.featurecleaning)

There are different types of date format and variations that are visible. Lets try find out various formats e.g.
19-06-2007 <br>
19-06-2007 00:00:00 <br>
19.6.07 <br>
19/06/2007 <br>
2007-19-06 <br>
etc<br>
There are extra white spaces in between dates

In [48]:
df['location'].value_counts()

Karve Road          3894
Pimpri Chinchwad    2016
Nal Stop            1282
Bhosari             1227
Swargate            1191
Pimpri               254
Name: location, dtype: int64

In [57]:
# for date analysis, first lets split the dataframe loations wise. for our analysis lets take Karve road readings
df_location_group = df.groupby('location')

In [58]:
df_karveroad = df_location_group.get_group('Karve Road').reset_index()

In [61]:
df_karveroad.drop(['index','sr.no.'],axis=1,inplace=True)

In [62]:
df_karveroad

date  so2  nox  rspm  spm  co2  aqi    location
0     19-06-2007    6   46    49    0    0   58  Karve Road
1     20-06-2007    6   43    29    0    0   54  Karve Road
2     21-06-2007    5   43    26    0    0   54  Karve Road
3     23-06-2007    7   46    79    0    0   79  Karve Road
4     24-06-2007    5   45    19    0    0   56  Karve Road
...          ...  ...  ...   ...  ...  ...  ...         ...
3889  20-03-2019   11   23   125    0    0  117  Karve Road
3890  21-03-2019   11   23    97    0    0   97  Karve Road
3891  22-03-2019   14   29   110    0    0  107  Karve Road
3892  23-03-2019   12   26   140    0    0  127  Karve Road
3893  24-03-2019   14   29   153    0    0  135  Karve Road

[3894 rows x 8 columns]

In [66]:
import sys
def dateformatfinder(myvalue):
    '''
    This function takes in date value, converts to string. Then removes time, extra white spaces and convers to 
    list of date, month and year in which ever format given to it.
    '''
    try:
        #convert to string
        curr = str(myvalue)
        #strip the time part if present
        curr1 = re.sub(r'\s[0-9]+:[0-9]+:[0-9]+','',curr)
        #remove blank spaces and replace anything which is not date with forward slash
        curr2 = re.sub(r'\s+','',curr1)
        curr3 = re.sub(r'[-.]','/',curr2)
        return curr3#[int(i) for i in curr3.split('/')]
    except:
        print(sys.exc_info())

In [67]:
dates = df_karveroad['date'].apply(dateformatfinder)

In [68]:
df_karveroad['date'] = dates

In [69]:
df_karveroad

date  so2  nox  rspm  spm  co2  aqi    location
0     19/06/2007    6   46    49    0    0   58  Karve Road
1     20/06/2007    6   43    29    0    0   54  Karve Road
2     21/06/2007    5   43    26    0    0   54  Karve Road
3     23/06/2007    7   46    79    0    0   79  Karve Road
4     24/06/2007    5   45    19    0    0   56  Karve Road
...          ...  ...  ...   ...  ...  ...  ...         ...
3889  20/03/2019   11   23   125    0    0  117  Karve Road
3890  21/03/2019   11   23    97    0    0   97  Karve Road
3891  22/03/2019   14   29   110    0    0  107  Karve Road
3892  23/03/2019   12   26   140    0    0  127  Karve Road
3893  24/03/2019   14   29   153    0    0  135  Karve Road

[3894 rows x 8 columns]

In [70]:
dates = np.array(dates.to_list())

In [71]:
 dates = np.array(list(np.char.split(dates,'/'))).astype(int)

In [74]:
# Now lets try to recognize pattern changes by subtracting date with previous date.
# if there is no format change maximum difference would be 30 for date, 11 for month, 1 for year. so in total 42
# if the substraction is greater than 42 we will detect a format change.
# This needs to happen location wise
date_diff = np.abs(dates[1:,:] - dates[:-1,:])

In [94]:
date_change = np.where(date_diff.sum(axis=1).reshape(date_diff.shape[0],1)>41)[0]
#at all the rows the date format is geeting changed
date_change.shape #there are 281 changes in formatting

(281,)

In [113]:
#Let us use the fancy indexing of Numpy to understand these format changes
np.where((dates[date_change+1][:,0]>1000) & (dates[date_change+1][:,1]<32) & (dates[date_change+1][:,2]<13))[0]

array([  0,   2,   4,   6,   8,  10,  12,  14,  16,  18,  20,  22,  24,
        26,  28,  30,  32,  34,  36,  38,  40,  42,  44,  46,  48,  50,
        52,  54,  56,  58,  60,  62,  64,  66,  68,  70,  72,  74,  76,
        78,  80,  82,  84,  86,  88,  90,  92,  94,  96,  98, 100, 102,
       104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128,
       130, 132, 134, 136, 138, 140, 142, 144, 146, 149, 151, 153, 155,
       157, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181,
       183, 185, 187, 189, 191, 193, 195, 197, 199, 201, 203, 205, 207,
       209, 211, 213, 215, 217, 219, 221, 223, 225, 227, 229, 231, 233,
       235, 237, 239, 241, 243, 245, 247, 249, 251, 253, 255, 257, 259,
       261, 263, 265, 267, 269, 271, 273, 275, 277, 279])

Out of 281 date changes, 140 are in probably in format %Y/%d/%m

In [104]:
#Let us use the fancy indexing of Numpy to understand these format changes
np.where((dates[date_change+1][:,0]>1000) & (dates[date_change+1][:,2]>12))[0].shape

(0,)

Out of this 140 non are having date in first column. So it is safe to say that these 140 are in format %Y/%d/%m

In [114]:
#Let us use the fancy indexing of Numpy to understand these format changes
np.where((dates[date_change+1][:,2]>1000) & (dates[date_change+1][:,1]<32) & (dates[date_change+1][:,0]<13))[0]

array([], dtype=int64)

There seems to be nothing in %m/%d/%Y and considering the Indian date system it generally goes %d/%m/%Y so lets see if it is in that format

In [115]:
#Let us use the fancy indexing of Numpy to understand these format changes
np.where((dates[date_change+1][:,2]>1000) & (dates[date_change+1][:,0]<32) & (dates[date_change+1][:,1]<13))[0]

array([  1,   3,   5,   7,   9,  11,  13,  15,  17,  19,  21,  23,  25,
        27,  29,  31,  33,  35,  37,  39,  41,  43,  45,  47,  49,  51,
        53,  55,  57,  59,  61,  63,  65,  67,  69,  71,  73,  75,  77,
        79,  81,  83,  85,  87,  89,  91,  93,  95,  97,  99, 101, 103,
       105, 107, 109, 111, 113, 115, 117, 119, 121, 123, 125, 127, 129,
       131, 133, 135, 137, 139, 141, 143, 145, 148, 150, 152, 154, 156,
       158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182,
       184, 186, 188, 190, 192, 194, 196, 198, 200, 202, 204, 206, 208,
       210, 212, 214, 216, 218, 220, 222, 224, 226, 228, 230, 232, 234,
       236, 238, 240, 242, 244, 246, 248, 250, 252, 254, 256, 258, 260,
       262, 264, 266, 268, 270, 272, 274, 276, 278, 280])

It is safe to say that these 140 are in format %d/%m/%Y

In [116]:
#lets try to find one outlier
#Let us use the fancy indexing of Numpy to understand these format changes
np.where((dates[date_change+1][:,2]<20) & (dates[date_change+1][:,0]<32) & (dates[date_change+1][:,1]<13))[0]

array([147])

One entry seems to be in format with short year

In [112]:
#lets try to find one outlier
#Let us use the fancy indexing of Numpy to understand these format changes
np.where((dates[date_change+1][:,2]<20) & (dates[date_change+1][:,0]<32) & (dates[date_change+1][:,1]<13))

(array([147]),)

In [127]:
(dates[date_change+1])[147]

array([13,  9, 13])

In [128]:
# Lets cover converting date formta 1 - %Y/%d/%m
date1 = pd.to_datetime(df_karveroad['date'],errors='coerce',format='%Y/%d/%m')

In [130]:
# Lets cover converting date formta 2 - %d/%m/%Y
date2 = pd.to_datetime(df_karveroad['date'],errors='coerce',format='%d/%m/%Y')

In [131]:
# Lets cover converting date formta 2 - %d/%m/%y
date3 = pd.to_datetime(df_karveroad['date'],errors='coerce',format='%d/%m/%y')

In [132]:
date_final = date1.fillna(date2).fillna(date3)

In [133]:
df_karveroad['date'] = date_final

In [134]:
df_karveroad[df_karveroad['date'].isnull()]
# All the dates are converted properly

Empty DataFrame
Columns: [date, so2, nox, rspm, spm, co2, aqi, location]
Index: []

In [135]:
df_karveroad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3894 entries, 0 to 3893
Data columns (total 8 columns):
date        3894 non-null datetime64[ns]
so2         3894 non-null int64
nox         3894 non-null int64
rspm        3894 non-null int64
spm         3894 non-null int64
co2         3894 non-null int64
aqi         3894 non-null int64
location    3894 non-null object
dtypes: datetime64[ns](1), int64(6), object(1)
memory usage: 243.5+ KB


In [136]:
df_karveroad.to_csv('clean_karvenagar_all.csv')

In [163]:
def dateformatchanger(df,column_date,export_path,columns_to_drop):
    date_series = df[column_date].apply(dateformatfinder)
    # Lets cover converting date formta 1 - %Y/%d/%m
    date1 = pd.to_datetime(date_series,errors='coerce',format='%Y/%d/%m')
    date2 = pd.to_datetime(date_series,errors='coerce',format='%d/%m/%Y')
    date3 = pd.to_datetime(date_series,errors='coerce',format='%d/%m/%y')
    df[column_date] = date1.fillna(date2).fillna(date3)
    df.drop(columns_to_drop,axis=1,inplace=True)
    print(df.info())
    df.to_csv(export_path)

Karve Road          3894
Pimpri Chinchwad    2016
Nal Stop            1282
Bhosari             1227
Swargate            1191
Pimpri               254

In [164]:
df_pcmc = df_location_group.get_group('Pimpri Chinchwad').reset_index()
df_ns = df_location_group.get_group('Nal Stop').reset_index()
df_bhosari = df_location_group.get_group('Bhosari').reset_index()
df_Swargate = df_location_group.get_group('Swargate').reset_index()
df_pimpri = df_location_group.get_group('Pimpri').reset_index()

In [166]:
dateformatchanger(df_pcmc,'date','clean_pcmc_all.csv',['sr.no.','index'])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2016 entries, 0 to 2015
Data columns (total 8 columns):
date        2016 non-null datetime64[ns]
so2         2016 non-null int64
nox         2016 non-null int64
rspm        2016 non-null int64
spm         2016 non-null int64
co2         2016 non-null int64
aqi         2016 non-null int64
location    2016 non-null object
dtypes: datetime64[ns](1), int64(6), object(1)
memory usage: 126.1+ KB
None


In [167]:
dateformatchanger(df_ns,'date','clean_ns_all.csv',['sr.no.','index'])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1282 entries, 0 to 1281
Data columns (total 8 columns):
date        1282 non-null datetime64[ns]
so2         1282 non-null int64
nox         1282 non-null int64
rspm        1282 non-null int64
spm         1282 non-null int64
co2         1282 non-null int64
aqi         1282 non-null int64
location    1282 non-null object
dtypes: datetime64[ns](1), int64(6), object(1)
memory usage: 80.2+ KB
None


In [168]:
dateformatchanger(df_bhosari,'date','clean_bhosari_all.csv',['sr.no.','index'])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1227 entries, 0 to 1226
Data columns (total 8 columns):
date        1227 non-null datetime64[ns]
so2         1227 non-null int64
nox         1227 non-null int64
rspm        1227 non-null int64
spm         1227 non-null int64
co2         1227 non-null int64
aqi         1227 non-null int64
location    1227 non-null object
dtypes: datetime64[ns](1), int64(6), object(1)
memory usage: 76.8+ KB
None


In [169]:
dateformatchanger(df_Swargate,'date','clean_swargate_all.csv',['sr.no.','index'])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1191 entries, 0 to 1190
Data columns (total 8 columns):
date        1191 non-null datetime64[ns]
so2         1191 non-null int64
nox         1191 non-null int64
rspm        1191 non-null int64
spm         1191 non-null int64
co2         1191 non-null int64
aqi         1191 non-null int64
location    1191 non-null object
dtypes: datetime64[ns](1), int64(6), object(1)
memory usage: 74.6+ KB
None


In [170]:
dateformatchanger(df_pimpri,'date','clean_pimpri_all.csv',['sr.no.','index'])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 254 entries, 0 to 253
Data columns (total 8 columns):
date        254 non-null datetime64[ns]
so2         254 non-null int64
nox         254 non-null int64
rspm        254 non-null int64
spm         254 non-null int64
co2         254 non-null int64
aqi         254 non-null int64
location    254 non-null object
dtypes: datetime64[ns](1), int64(6), object(1)
memory usage: 16.0+ KB
None
